src/shared/keyboard-navigation.js

```diff
@@ -0,0 +1,197 @@
+  } = {}
+) {
+  useEffect(() => {
+    if (!containerRef.current) {
+      throw new Error('Container ref not set');
+    }
+    const container = containerRef.current;
+
+    const getNavigableElements = () => {
+      const elements = /** @type {HTMLElement[]} */ (
+        Array.from(container.querySelectorAll(selector))
+      );
+      return elements.filter(
+        el => isElementVisible(el) && !isElementDisabled(el)
+      );
+    };
+
+    /**
+     * Update the `tabindex` attribute of navigable elements.
+     *
+     * Exactly one element will have `tabindex=0` and all others will have
+     * `tabindex=1`.
+     *
+     * @param {HTMLElement[]} elements
+     * @param {number} currentIndex - Index of element in `elements` to make current.
+     *   Defaults to the current element if there is one, or the first element
+     *   otherwise.
+     * @param {boolean} setFocus - Whether to focus the current element
+     */
+    const updateTabIndexes = (
+      elements = getNavigableElements(),
+      currentIndex = -1,
+      setFocus = false
+    ) => {
+      if (currentIndex < 0) {
+        currentIndex = elements.findIndex(el => el.tabIndex === 0);
+        if (currentIndex < 0) {
+          currentIndex = 0;
+        }
+      }
+
+      for (let [index, element] of elements.entries()) {
+        element.tabIndex = index === currentIndex ? 0 : -1;
+        if (index === currentIndex && setFocus) {
+          element.focus();
+        }
+      }
+    };
+
+    /** @param {KeyboardEvent} event */
+    const onKeyDown = event => {
+      const elements = getNavigableElements();
+      let currentIndex = elements.findIndex(item => item.tabIndex === 0);
+
+      let handled = false;
+      if (
+        (horizontal && event.key === 'ArrowLeft') ||
+        (vertical && event.key === 'ArrowUp')
+      ) {
+        if (currentIndex === 0) {
+          currentIndex = elements.length - 1;
+        } else {
+          --currentIndex;
+        }
+        handled = true;
+      } else if (
+        (horizontal && event.key === 'ArrowRight') ||
+        (vertical && event.key === 'ArrowDown')
+      ) {
+        if (currentIndex === elements.length - 1) {
+          currentIndex = 0;
+        } else {
+          ++currentIndex;
+        }
+        handled = true;
+      } else if (event.key === 'Home') {
+        currentIndex = 0;
+        handled = true;
+      } else if (event.key === 'End') {
+        currentIndex = elements.length - 1;
+        handled = true;
+      }
+
+      if (!handled) {
+        return;
+      }
+
+      updateTabIndexes(elements, currentIndex, true);
+
+      event.preventDefault();
+      event.stopPropagation();
+    };
+
+    updateTabIndexes(getNavigableElements(), 0, autofocus);
+
+    const listeners = new ListenerCollection();
+
+    // Set an element as current when it gains focus. In Safari this event
+    // may not be received if the element immediately loses focus after it
+    // is triggered.
```

Comment:
In the case of MarkdownEditor, the text input field is re-focused as soon as the user clicks on one of the toolbar buttons. In Chrome and Firefox the toolbar container receives a `focusin` event, in Safari it does not.

Reply:
If the `focusin` event isn't received, `tabindex` doesn't update.

In the context of MarkdownEditor, suppose the "Bold" button in the toolbar initially has focus and the user _clicks_ on the "Italic" button. The focus will then transfer to the editor. In Chrome and Firefox, if the user back-tabs to the toolbar, the Italic button will receive focus. In Safari however, the Bold button will receive focus.

I don't think this is a huge issue because if the user selects an item from the toolbar using the keyboard instead of the mouse, the tabindex will get updated as the user selects an item using the left/right arrows.